cosmpy/aerial/contract/__init__.py

```diff
@@ -0,0 +1,250 @@
+# -*- coding: utf-8 -*-
+# ------------------------------------------------------------------------------
+#
+#   Copyright 2018-2021 Fetch.AI Limited
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+
+import json
+from datetime import datetime
+from typing import Any, Optional
+
+from cosmpy.aerial import LedgerClient
+from cosmpy.aerial.contract.cosmwasm import (
+    create_cosmwasm_execute_msg,
+    create_cosmwasm_instantiate_msg,
+    create_cosmwasm_store_code_msg,
+)
+from cosmpy.aerial.tx import SigningCfg, Transaction
+from cosmpy.aerial.tx_helpers import SubmittedTx
+from cosmpy.crypto.address import Address
+from cosmpy.crypto.hashfuncs import sha256
+from cosmpy.crypto.keypairs import PrivateKey
+from cosmpy.protos.cosmos.base.query.v1beta1.pagination_pb2 import PageRequest
+from cosmpy.protos.cosmwasm.wasm.v1.query_pb2 import (
+    QueryCodesRequest,
+    QuerySmartContractStateRequest,
+)
+
+
+def _compute_digest(path: str) -> bytes:
+    with open(path, "rb") as input_file:
+        return sha256(input_file.read())
+
+
+def _generate_label(digest: bytes) -> str:
+    now = datetime.utcnow()
+    return f"{digest.hex()[:14]}-{now.strftime('%Y%m%d%H%M%S')}"
+
+
+class LedgerContract:
+    def __init__(
+        self, path: str, client: LedgerClient, address: Optional[Address] = None
+    ):
+        self._path = path
+        self._client = client
+        self._digest = _compute_digest(self._path)
+        self._code_id = self._find_contract_id_by_digest(self._digest)
+        self._address = address
+
+    @property
+    def path(self) -> str:
+        return self._path
+
+    @property
+    def digest(self) -> bytes:
+        return self._digest
+
+    @property
+    def code_id(self) -> Optional[int]:
+        return self._code_id
+
+    @property
+    def address(self) -> Optional[Address]:
+        return self._address
+
+    def store(self, key: PrivateKey, gas_limit: Optional[int] = None) -> int:
+        sender_address = Address(key)
+
+        # query the account information for the sender
+        account = self._client.query_account(sender_address)
+
+        # estimate the fee required for this transaction
+        gas_limit = (
+            gas_limit or 2000000
```

Comment:
We are also able to query the gas limit from the node. It can also return -1 when gas is not limited, so we should replace it with some arbitrary value.

Reply:
We are also able to query the gas limit from the node. It can also return -1 when gas is not limited, so we should replace it with some arbitrary value.

Reply:
yup indeed, that's the best option. Would still love to see some manual override options (as fetchd, for all gas settings, account sequence....) just in case one does not want to hit the network so much before broadcasting